### PR TITLE
add nn4sys benchmark

### PR DESCRIPTION
**Pull request by Wei-TianHao**

### CODE REVIEW

**Comment by stanleybak** [comment]
The sum of the timeouts in the instances.csv file (35606 seconds) exceeds the allowed six hour timeout. Please adjust.

Also: while not required, I suggest to get rid of many of the instances in the .csv file. For example, many of the files are individual cases of the disjunction. Let the first 10 benchmarks be these easy instances, and then for the rest just use disjunctions. Just an opinion though.

**Comment by stanleybak** [comment]
Another issue: the `.onnx.gz` files are compressed `.tar` files. Just compress the `.onnx` file directly (no need to tar it): `gzip -k network.onnx`

**Comment by Wei-TianHao** [comment]
Hi Stanley,
Sorry I mistook the total time. Thanks for your advice. I have removed most individual cases to meet the time requirement and recompressed network files with gzip.

**Comment by stanleybak** [comment]
`nn4sys/specs/normal_100_101.vnnlib` fails to parse:

`(assert (or (and (>= Y_0 29074)((<= Y_0 30291)))(and (>= Y_0 -1)(<= Y_0 1))))`

The string `((<= Y_0 30291))` looks incorrect.

**Comment by Wei-TianHao** [comment]
I removed redundant brackets. Thanks.

**Comment by stanleybak** [comment]
The spec file `normal_100_1_2_3_4_5.vnnlib` has a line that fails to parse:

`(assert (or (and (>= X_0 135845.0)(<= X_0 216892.0)(> Y_0 1419.0))(and (>= X_0 224027.0)(<= X_0 247495.0)(> Y_0 2942.0))(and (>= X_0 252460.0)(<= X_0 267771.0)(> Y_0 4464.0))(and (>= X_0 271062.0)(<= X_0 282649.0)(> Y_0 5986.0))(and (>= X_0 285308.0)(<= X_0 294745.0)(> Y_0 7508.0))))`

Specifically, strict inequalities like `>` are not allowed by the spec. You can try running things with `simple_adversarial_generator` first to see if things parse correctly.

**Comment by stanleybak** [comment]
Also, the first 9 benchmarks are violated using random tests... not sure if this was intentional.

**Comment by Wei-TianHao** [comment]
Sorry for the trouble. The parser I wrote is too simple, it didn't find these problems. I'll double check it. I apologize for the inconvenience.

**Comment by huanzhang12** [comment]
@Wei-TianHao I found that in `nn4sys_instances.csv` there are some instances with only 1 or 2 seconds timeout? That is probably too short - some tools may not even able to fully read the model file into memory within 1 or 2 seconds (some unzipped onnx files are over 100MB), and some large python packages like Tensorflow/PyTorch can take a few seconds to import. Maybe we want to set the minimum timeout threshold to something like 10 seconds?

**Comment by Wei-TianHao** [comment]
Hi @stanleybak ,

When I run `python ../src/agen/randgen.py test_sat.onnx test_prop.vnnlib out.txt`. It also returns `violated`. It seems you changed `res = 'sat'` to  `res = 'violated'` [here](https://github.com/stanleybak/simple_adversarial_generator/blob/fb79c29e0bfb87ef44a1536fd1ae1ff93397aacc/src/agen/randgen.py#L85). I guess that's why it thinks all the benchmarks are violated. Am I understanding it right? Thank you!

**Comment by Wei-TianHao** [comment]
> @Wei-TianHao I found that in `nn4sys_instances.csv` there are some instances with only 1 or 2 seconds timeout? That is probably too short - some tools may not even able to fully read the model file into memory within 1 or 2 seconds (some unzipped onnx files are over 100MB), and some large python packages like Tensorflow/PyTorch can take a few seconds to import. Maybe we want to set the minimum timeout threshold to something like 10 seconds?

That's a good point. We will consider that. Thank you.

**Comment by mnmueller** [comment]
@Wei-TianHao My understanding is that the specs effectively define a permissible counterexample. So if there is an assignment satisfying all asserts in the spec file, this is a counterexample. Hence "sat"="violated". Maybe you can check if this is inline with the intention of the specification.

**Comment by stanleybak** [comment]
@Wei-TianHao yes I agree using timeouts more than a few seconds is better. @mnmueller and you are also right that `violated` means what `sat` used to mean before (we might change this next year, but it was in the rules this time so that's why I changed it in the `randgen.py`). The `vnnlib` file encodes a counterexample, and then the tool checks if all the constraints can be met at the same time, which shows the existence of a counterexample, so therefore the desired property is violated.

**Comment by Wei-TianHao** [comment]
@mnmueller @stanleybak Thanks for your help! I updated the specs. I found that some of the violations are caused by precision errors (Julia and python produce different outputs under the same condition), so I also changed network weights to `double`. The benchmark passed `simple_adversarial_generator` parsing.

**Comment by Wei-TianHao** [comment]
@huanzhang12 We relaxed the timeout to make sure tools have time to set up. Thanks for your advice!

**Comment by stanleybak** [comment]
Were the inputs normalized during training? I think using the normalized input constraints would help fix some of the floating point issues, as right now the inputs are fairly large numbers which may lose precision in the network. I'm also not sure how many tools will respect the double vs float semantics in onnx files (I needed to add support to this for the `simple_adversarial_generator`, for example). Anyways, since things run now I'll merge and if you want to do updates later we can do another pull request.

**Comment by Wei-TianHao** [comment]
Yes. The first layer of the network is a normalization layer which compresses the input. It's the special network architecture that amplifies the precision error.
